sale_coupon_multi_use/models/sale_order.py

```diff
@@ -1,32 +1,134 @@
 
     _inherit = "sale.order"
 
-    # TODO: implement multi-use coupon relation on sale.order ->
-    # applied_coupon_ids is o2m, which removes coupon if same coupon
-    # is applied on another SO. But for multi-use coupon, it must be
-    # able to keep same coupon on multiple sale orders.
-    # TODO: action_draft seems buggy, because it does not reset back
-    # coupons to new state (you can abuse coupon this way and use normal
-    # coupon multiple times).
+    coupon_multi_use_ids = fields.Many2many(
+        "sale.coupon",
+        "sale_order_coupon_multi_rel",
+        "sale_id",
+        "coupon_id",
+        string="Multi Use Coupons",
+        copy=False,
+        readonly=True,
+    )
+
+    def _get_multi_use_coupons_to_handle(self, filter_func=None):
+        self.ensure_one()
+        coupons = self.coupon_multi_use_ids - self.applied_coupon_ids
+        if filter_func:
+            return filter_func(coupons)
+        return coupons
+
+    def _trigger_multi_use_coupons(self, state, filter_func=None):
+        self.ensure_one()
+        # NOTE. This method must be called with ORDER_CTX_KEY on order.
+        multi_use_coupons = self._get_multi_use_coupons_to_handle()
+        if multi_use_coupons:
+            multi_use_coupons.write({"state": state})
+        return multi_use_coupons
 
     def action_confirm(self):
         """Extend to pass coupon_sale_order context."""
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            # Mimic same behavior as for single-user coupon.
+            order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_confirm()
+            order._trigger_multi_use_coupons("used")
         return True
 
     def action_cancel(self):
         """Extend to pass coupon_sale_order context."""
-        # NOTE. Can't rely on coupon/SO relation, because
-        # applied_coupon_ids is o2m field, meaning same coupon used on
-        # another SO, would make lose relation on previous SO.
         for order in self:
-            order = order.with_context(coupon_sale_order=order)
+            order = order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_cancel()
+            order._trigger_multi_use_coupons("new")
+        return True
+
+    def action_draft(self):
+        """Extend to pass coupon_sale_order context."""
+        super().action_draft()
+        for order in self:
+            # TODO: refactor this to use same approach as for
+            # action_confirm and action_cancel if Odoo fixes their bug.
+            # Now single use coupons are not reset back if you put SO to
+            # draft state.
+            coupons_multi_use = order.coupon_multi_use_ids
+            if coupons_multi_use:
+                coupons_multi_use.with_context(**{ORDER_CTX_KEY: order}).write(
+                    {"state": "used"}
+                )
```

Comment:
No need to test `coupons_multi_use`.
Let Odoo manage it in `write` function.
```suggestion
            order.with_context(**{ORDER_CTX_KEY: order}).coupon_multi_use_ids.write(
                {"state": "used"}
            )
```

Reply:
It will call write with empty recordset, which is ugly.